File: app/src/main/java/com/androidizate/vale4/MainActivity.java
package com.androidizate.vale4;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    int puntajeNosotros = 0;
    int puntajeEllos = 0;

    TextView scoreNosotros;
    Button envidoNosotros;
    Button realEnvidoNosotros;
    Button faltaEnvidoNosotros;
    Button trucoNosotros;
    Button reTrucooNosotros;
    Button valeCuatroNosotros;
    Button puntoNosotros;
    Button sumarPuntoNosotros;
    Button restarPuntoNosotros;

    TextView scoreEllos;
    Button envidoEllos;
    Button realEnvidoEllos;
    Button faltaEnvidoEllos;
    Button trucoEllos;
    Button reTrucooEllos;
    Button valeCuatroEllos;
    Button puntoEllos;
    Button sumarPuntoEllos;
    Button restarPuntoEllos;

    Button borrarPuntajes;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initNosotros();
        initEllos();

        borrarPuntajes = (Button) findViewById(R.id.btn_reset);
        borrarPuntajes.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                resetearPuntaje();
            }
        });
    }

    private void initEllos() {
        scoreEllos = (TextView) findViewById(R.id.score_ellos);

        envidoEllos = (Button) findViewById(R.id.btn_envido_ellos);
        envidoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarEnvidoEllos();
            }
        });

        realEnvidoEllos = (Button) findViewById(R.id.btn_real_envido_ellos);
        realEnvidoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarRealEnvidoEllos();
            }
        });

        faltaEnvidoEllos = (Button) findViewById(R.id.btn_falta_envido_ellos);
        faltaEnvidoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarFaltaEnvidoEllos();
            }
        });

        trucoEllos = (Button) findViewById(R.id.btn_truco_ellos);
        trucoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarTrucoEllos();
            }
        });

        reTrucooEllos = (Button) findViewById(R.id.btn_re_truco_ellos);
        reTrucooEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarReTrucoEllos();
            }
        });

        valeCuatroEllos = (Button) findViewById(R.id.btn_vale_cuatro_ellos);
        valeCuatroEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarValeCuatroEllos();
            }
        });

        puntoEllos = (Button) findViewById(R.id.btn_punto_ellos);
        puntoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarPuntoEllos();
            }
        });

        sumarPuntoEllos = (Button) findViewById(R.id.btn_sumar_uno_ellos);
        sumarPuntoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarPuntoEllos();
            }
        });

        restarPuntoEllos = (Button) findViewById(R.id.btn_restar_uno_ellos);
        restarPuntoEllos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                restarPuntoEllos();
            }
        });
    }

    private void initNosotros() {
        scoreNosotros = (TextView) findViewById(R.id.score_nosotros);

        envidoNosotros = (Button) findViewById(R.id.btn_envido_nosotros);
        envidoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarEnvidoNosotros();
            }
        });

        realEnvidoNosotros = (Button) findViewById(R.id.btn_real_envido_nosotros);
        realEnvidoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarRealEnvidoNosotros();
            }
        });

        faltaEnvidoNosotros = (Button) findViewById(R.id.btn_falta_envido_nosotros);
        faltaEnvidoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarFaltaEnvidoNosotros();
            }
        });

        trucoNosotros = (Button) findViewById(R.id.btn_truco_nosotros);
        trucoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarTrucoNosotros();
            }
        });

        reTrucooNosotros = (Button) findViewById(R.id.btn_re_truco_nosotros);
        reTrucooNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarReTrucoNosotros();
            }
        });

        valeCuatroNosotros = (Button) findViewById(R.id.btn_vale_cuatro_nosotros);
        valeCuatroNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarValeCuatroNosotros();
            }
        });

        puntoNosotros = (Button) findViewById(R.id.btn_punto_nosotros);
        puntoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarPuntoNosotros();
            }
        });

        sumarPuntoNosotros = (Button) findViewById(R.id.btn_sumar_uno_nosotros);
        sumarPuntoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sumarPuntoNosotros();
            }
        });

        restarPuntoNosotros = (Button) findViewById(R.id.btn_restar_uno_nosotros);
        restarPuntoNosotros.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                restarPuntoNosotros();
            }
        });
    }

    private void restarPuntoEllos() {
        if (puntajeEllosMayorACero()) {
            puntajeEllos -= 1;
            cambiarScoreEllos();
        }
    }

    private void restarPuntoNosotros() {
        if (puntajeNosotrosMayorACero()) {
            puntajeNosotros -= 1;
            cambiarScoreNosotros();
        }
    }

    private void sumarPuntoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 1;
            cambiarScoreEllos();
        }
    }

    private void sumarPuntoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 1;
            cambiarScoreNosotros();
        }
    }

    private void sumarEnvidoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 2;
            cambiarScoreEllos();
        }
    }

    private void sumarEnvidoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 2;
            cambiarScoreNosotros();
        }
    }

    private void sumarRealEnvidoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 3;
            cambiarScoreEllos();
        }
    }

    private void sumarRealEnvidoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 3;
            cambiarScoreNosotros();
        }
    }

    private void sumarFaltaEnvidoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 30 - puntajeNosotros;
            cambiarScoreEllos();
        }
    }

    private void sumarFaltaEnvidoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 30 - puntajeEllos;
            cambiarScoreNosotros();
        }
    }

    private void sumarTrucoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 2;
            cambiarScoreEllos();
        }
    }

    private void sumarTrucoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 2;
            cambiarScoreNosotros();
        }
    }

    private void sumarReTrucoEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 3;
            cambiarScoreEllos();
        }
    }

    private void sumarReTrucoNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 3;
            cambiarScoreNosotros();
        }
    }

    private void sumarValeCuatroEllos() {
        if (puntajeEllosMenorATreinta()) {
            puntajeEllos += 4;
            cambiarScoreEllos();
        }
    }

    private void sumarValeCuatroNosotros() {
        if (puntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 4;
            cambiarScoreNosotros();
        }
    }

    private void cambiarScoreEllos() {
        if (puntajeEllos >= 30) {
            puntajeEllos = 30;
            new AlertDialog.Builder(this)
                    .setTitle("Fin de la partida")
                    .setNeutralButton("Ok", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            puntajeEllos = 0;
                            puntajeNosotros = 0;
                            cambiarScoreEllos();
                            cambiarScoreNosotros();
                            dialogInterface.dismiss();
                        }
                    })
                    .setMessage("Ganaron ELL@S!!")
                    .show();
        }
        scoreEllos.setText(String.valueOf(puntajeEllos));
    }

    private void cambiarScoreNosotros() {
        if (puntajeNosotros >= 30) {
            puntajeNosotros = 30;
            new AlertDialog.Builder(this)
                    .setTitle("Fin de la partida")
                    .setNeutralButton("Ok", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            puntajeEllos = 0;
                            puntajeNosotros = 0;
                            cambiarScoreEllos();
                            cambiarScoreNosotros();
                            dialogInterface.dismiss();
                        }
                    })
                    .setMessage("Ganamos NOSOTR@S!!")
                    .show();
        }
        scoreNosotros.setText(String.valueOf(puntajeNosotros));
    }

    private boolean puntajeEllosMayorACero() {
        return puntajeEllos > 0;
    }

    private boolean puntajeNosotrosMayorACero() {
        return puntajeNosotros > 0;
    }

    private boolean puntajeEllosMenorATreinta() {
        return puntajeEllos < 30;
    }

    private boolean puntajeNosotrosMenorATreinta() {
        return puntajeNosotros < 30;
    }

    private void resetearPuntaje() {
        new AlertDialog.Builder(this)
                .setTitle("Cuidado")
                .setPositiveButton("Si", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        puntajeEllos = 0;
                        puntajeNosotros = 0;
                        cambiarScoreEllos();
                        cambiarScoreNosotros();
                        dialogInterface.dismiss();
                    }
                })
                .setNegativeButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        dialogInterface.dismiss();
                    }
                })
                .setMessage("Está seguro que desea borrar el puntaje y empezar nuevamente?")
                .show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/androidizate/vale4/MainActivity.kt
package com.androidizate.vale4

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    var puntajeNosotros = 0
    var puntajeEllos = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initNosotros()
        initEllos()
        btn_reset.setOnClickListener { resetearPuntaje() }
    }

    private fun initEllos() {
        btn_envido_ellos.setOnClickListener { sumarEnvidoEllos() }
        btn_real_envido_ellos.setOnClickListener { sumarRealEnvidoEllos() }
        btn_falta_envido_ellos.setOnClickListener { sumarFaltaEnvidoEllos() }
        btn_truco_ellos.setOnClickListener { sumarTrucoEllos() }
        btn_re_truco_ellos.setOnClickListener { sumarReTrucoEllos() }
        btn_vale_cuatro_ellos.setOnClickListener { sumarValeCuatroEllos() }
        btn_punto_ellos.setOnClickListener { sumarUnoEllos() }
        btn_sumar_uno_ellos.setOnClickListener { sumarUnoEllos() }
        btn_restar_uno_ellos.setOnClickListener { restarUnoEllos() }
    }

    private fun initNosotros() {
        btn_envido_nosotros.setOnClickListener { sumarEnvidoNosotros() }
        btn_real_envido_nosotros.setOnClickListener { sumarRealEnvidoNosotros() }
        btn_falta_envido_nosotros.setOnClickListener { sumarFaltaEnvidoNosotros() }
        btn_truco_nosotros.setOnClickListener { sumarTrucoNosotros() }
        btn_re_truco_nosotros.setOnClickListener { sumarReTrucoNosotros() }
        btn_vale_cuatro_nosotros.setOnClickListener { sumarValeCuatroNosotros() }
        btn_punto_nosotros.setOnClickListener { sumarUnoNosotros() }
        btn_sumar_uno_nosotros.setOnClickListener { sumarUnoNosotros() }
        btn_restar_uno_nosotros.setOnClickListener { restarUnoNosotros() }
    }

    private fun restarUnoEllos() {
        if (isPuntajeEllosMayorACero()) {
            puntajeEllos -= 1
            cambiarScoreEllos()
        }
    }

    private fun restarUnoNosotros() {
        if (isPuntajeNosotrosMayorACero()) {
            puntajeNosotros -= 1
            cambiarScoreNosotros()
        }
    }

    private fun sumarUnoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 1
            cambiarScoreEllos()
        }
    }

    private fun sumarUnoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 1
            cambiarScoreNosotros()
        }
    }

    private fun sumarEnvidoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 2
            cambiarScoreEllos()
        }
    }

    private fun sumarEnvidoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 2
            cambiarScoreNosotros()
        }
    }

    private fun sumarRealEnvidoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 3
            cambiarScoreEllos()
        }
    }

    private fun sumarRealEnvidoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 3
            cambiarScoreNosotros()
        }
    }

    private fun sumarFaltaEnvidoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 30 - puntajeNosotros
            cambiarScoreEllos()
        }
    }

    private fun sumarFaltaEnvidoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 30 - puntajeEllos
            cambiarScoreNosotros()
        }
    }

    private fun sumarTrucoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 2
            cambiarScoreEllos()
        }
    }

    private fun sumarTrucoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 2
            cambiarScoreNosotros()
        }
    }

    private fun sumarReTrucoEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 3
            cambiarScoreEllos()
        }
    }

    private fun sumarReTrucoNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 3
            cambiarScoreNosotros()
        }
    }

    private fun sumarValeCuatroEllos() {
        if (isPuntajeEllosMenorATreinta()) {
            puntajeEllos += 4
            cambiarScoreEllos()
        }
    }

    private fun sumarValeCuatroNosotros() {
        if (isPuntajeNosotrosMenorATreinta()) {
            puntajeNosotros += 4
            cambiarScoreNosotros()
        }
    }

    private fun cambiarScoreEllos() {
        if (puntajeEllos >= 30) {
            puntajeEllos = 30
            AlertDialog.Builder(this)
                    .setTitle("Fin de la partida")
                    .setNeutralButton("Ok") { dialogInterface, _ ->
                        puntajeEllos = 0
                        puntajeNosotros = 0
                        cambiarScoreEllos()
                        cambiarScoreNosotros()
                        dialogInterface.dismiss()
                    }
                    .setMessage("Ganaron ELLES \n (se va a caer)")
                    .show()
        }
        score_ellos.text = puntajeEllos.toString()
    }

    private fun cambiarScoreNosotros() {
        if (puntajeNosotros >= 30) {
            puntajeNosotros = 30
            AlertDialog.Builder(this)
                    .setTitle("Fin de la partida")
                    .setNeutralButton("Ok") { dialogInterface, _ ->
                        puntajeEllos = 0
                        puntajeNosotros = 0
                        cambiarScoreEllos()
                        cambiarScoreNosotros()
                        dialogInterface.dismiss()
                    }
                    .setMessage("Ganamos NOSOTRES \n (se va a caer)")
                    .show()
        }
        score_nosotros.text = puntajeNosotros.toString()
    }

    private fun isPuntajeEllosMayorACero(): Boolean {
        return puntajeEllos > 0
    }

    private fun isPuntajeNosotrosMayorACero(): Boolean {
        return puntajeNosotros > 0
    }

    private fun isPuntajeEllosMenorATreinta(): Boolean {
        return puntajeEllos < 30
    }

    private fun isPuntajeNosotrosMenorATreinta(): Boolean {
        return puntajeNosotros < 30
    }

    private fun resetearPuntaje() {
        AlertDialog.Builder(this)
                .setTitle("Cuidado")
                .setPositiveButton("Si") { dialogInterface, _ ->
                    puntajeEllos = 0
                    puntajeNosotros = 0
                    cambiarScoreEllos()
                    cambiarScoreNosotros()
                    dialogInterface.dismiss()
                }
                .setNegativeButton("No") { dialogInterface, _ -> dialogInterface.dismiss() }
                .setMessage("Está seguro que desea borrar el puntaje y empezar nuevamente?")
                .show()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_settings) {
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}